cmake/modules/AddSwift.cmake

```diff
@@ -575,34 +575,51 @@ function(_add_swift_library_single target name)
       INSTALL_IN_COMPONENT "${SWIFTLIB_INSTALL_IN_COMPONENT}")
   add_swift_source_group("${SWIFTLIB_SINGLE_EXTERNAL_SOURCES}")
 
-  add_library("${target}" ${libkind}
-      ${SWIFT_SECTIONS_OBJECT_BEGIN}
-      ${SWIFTLIB_SINGLE_SOURCES}
-      ${SWIFTLIB_SINGLE_EXTERNAL_SOURCES}
-      ${SWIFT_SECTIONS_OBJECT_END})
-
-  # The section metadata objects are generated sources, and we need to tell CMake
-  # not to expect to find them prior to their generation.
-  if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux" OR
-     "${CMAKE_SYSTEM_NAME}" STREQUAL "FreeBSD")
-    if("${libkind}" STREQUAL "SHARED")
-      set_source_files_properties(${SWIFT_SECTIONS_OBJECT_BEGIN} PROPERTIES GENERATED 1)
-      set_source_files_properties(${SWIFT_SECTIONS_OBJECT_END} PROPERTIES GENERATED 1)
-      add_dependencies("${target}" section_magic)
+  # "All targets" is composed of a dynamic library target, a static library
+  # target, or both.
+  set(all_library_targets)
+
+  # First up, we conditionally add a dynamic library target.
+  if(SWIFTLIB_IS_STDLIB OR SWIFTLIB_IS_SDK_OVERLAY)
+    if(SWIFTLIB_SINGLE_IS_STDLIB AND SWIFT_BUILD_DYNAMIC_STDLIB)
+      set(add_dynamic_library_target TRUE)
+    elseif(SWIFTLIB_SINGLE_IS_SDK_OVERLAY AND SWIFT_BUILD_DYNAMIC_SDK_OVERLAY)
+      set(add_dynamic_library_target TRUE)
+    else()
+      set(add_dynamic_library_target FALSE)
     endif()
+  else()
+    set(add_dynamic_library_target TRUE)
   endif()
+  if(add_dynamic_library_target)
+    add_library("${target}" ${libkind}
```

Comment:
One wrinkle here -- this is not necessarily a dynamic library.  This `add_library()` call is also invoked for compiler libraries like `libSwiftAST`, which are static.